### PR TITLE
Add QSPI Driver to HAL ChibiOS

### PR DESCRIPTION
**Pull request by bugobliterator**
This PR is the "safe" part of the work towards adding support for external flash, the next part will include Bootloader changes, which can be bit more sensitive. This Just includes Low level driver and FlashIface library (which isn't even added to build for normal vehicles). https://github.com/ArduPilot/ardupilot/pull/17635

### CODE REVIEW

**Comment by bugobliterator** [comment]
Before:
```
Target         Text     Data  BSS     Total  
---------------------------------------------
bin/arduplane  1606480  2416  128856  1737752
```
After:
```
Target         Text     Data  BSS     Total  
---------------------------------------------
bin/arduplane  1606724  2420  128860  1738004
```

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+
+#define SFDP_HDR_NUM_PARAMS(x)              (SFDP_GET_BITS(x[1], 16, 19) + 1)
+#define SFDP_HDR_PARAM_REV(x)               SFDP_GET_BITS(x[1], 0, 15)
+#define SFDP_PARAM_ID(x)                    ((SFDP_GET_BITS(x[0], 0, 3) << 8) | SFDP_GET_BITS(x[1], 24, 31))
+#define SFDP_PARAM_DWORD_LEN(x)             SFDP_GET_BITS(x[0], 24, 31)
+#define SFDP_PARAM_POINTER(x)               SFDP_GET_BITS(x[1], 0, 23)
+
+#define SFDP_REV_1_6                        0x0106
+
+#define Debug(fmt, args ...)  do {hal.console->printf("FlashIface_JEDEC: " fmt "\n", ## args);} while(0)
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:68]: Small thing, but you are using a big chunk of your 50 chars on a long constant string

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 2]
```diff
@@ -0,0 +1,849 @@
...
+        }
+    }
+    // Reset Device involves trying to soft reset the chip
+    // as when system reboots the device might not have.
+    reset_device();
+
+    // Detecting Device involves trying to read Device ID and matching
+    // with what we expect. Along with extracting info from SFDP
+    if (!detect_device()) {
+        hal.console->printf("Failed to detect flash device: %s\n", supported_devices[_dev_list_idx].name);
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:90]: Seems kind of serious, maybe a GCS error?

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+
+    /* Single line N25Q_CMD_RESET_MEMORY command.*/
+    cmd.cmd   =  CMD_RESET_MEMORY;
+    cmd.cfg   =  AP_HAL::QSPI::CFG_CMD_MODE_ONE_LINE;
+    cmd.addr  =  0;
+    cmd.alt   =  0;
+    cmd.dummy =  0;
+    _dev->set_cmd_header(cmd);
+    _dev->transfer(nullptr, 0, nullptr, 0);
+
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:169]: At least on the winbond chips there is a prescribed delay between reset and doing anything else. You probably need some kind of delay in here

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 4]
```diff
@@ -0,0 +1,849 @@
...
+            Debug("Failed to read Parameter Table");
+            return false;
+        }
+
+        // Flash Memory details Ref. JESD216 6.4.5 6.4.14
+        if (SFDP_GET_BIT(param_table[1], 31)) {
+            Debug("Unsupported Flash Size");
+            return false;
+        }
+        _desc.flash_size = SFDP_GET_BITS(param_table[1], 0, 30)/8;
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:271]: A comment about these numbers

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 5]
```diff
@@ -0,0 +1,849 @@
...
+        _desc.page_size = 1UL<<SFDP_GET_BITS(param_table[10], 4, 7);
+        _desc.page_count = _desc.flash_size/_desc.page_size;
+        if (_desc.page_count == 0) {
+            Debug("Page size greater than flash size unsupported");
+            return false;
+        }
+
+        // Erase Flash Memory details Ref. JESD216 6.4.11 6.4.12
+        for (uint8_t i = 0; i < 4; i++) {
+            uint32_t size = 1UL<<SFDP_GET_BITS(param_table[7 + (i/2)], 0 + 16*(i%2), 7 + 16*(i%2));
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:281]: Having a hard time parsing the math - a comment would help

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 6]
```diff
@@ -0,0 +1,849 @@
...
+            uint8_t ins = SFDP_GET_BITS(param_table[7 + (i/2)], 8 + 16*(i%2), 15 + 16*(i%2));
+            if (size-1) {
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:283]: "> 0"

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 7]
```diff
@@ -0,0 +1,849 @@
...
+    // Enable 4-4-4 mode and test it by fetching Device ID
+    {
+        // Quad Mode Enable Sequence, Ref. JESD216 6.4.18
+        if (_desc.quad_mode_ins && !write_enable_called) {
+            if (!send_cmd(_desc.quad_mode_ins, false)) {
+                return false;
+            }
+        } else if (_desc.quad_mode_rmw_seq) {
+            write_enable(false);
+            if (!modify_reg(0x65, 0x61, (1<<7), 0, false)) {
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:421]: constants for these registers must be somewhere

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 8]
```diff
@@ -0,0 +1,849 @@
...
+    }
+    return true;
+}
+
+bool AP_FlashIface_JEDEC::read_reg(uint8_t read_ins, uint8_t &read_val, bool quad_mode)
+{
+    AP_HAL::QSPIDevice::CommandHeader cmd;
+    cmd.cmd = read_ins;
+    if (quad_mode) {
+        cmd.cfg    =  AP_HAL::QSPI::CFG_CMD_MODE_FOUR_LINES |
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:498]: Why not support dual SPI?

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 9]
```diff
@@ -0,0 +1,849 @@
...
+ * @param[out] delay_ms     Time to wait until next is_device_busy call
+ * @param[out] timeout_ms   Time after which the erase should be timedout,
+ *                          should be reset at every call.
+ * @return                  The operation status.
+ * @retval false            if the operation failed.
+ * @retval true             if the operation succeeded.
+ *
+ */
+bool AP_FlashIface_JEDEC::start_program_page(uint32_t page, uint8_t* data,
+        uint32_t &delay_us, uint32_t &timeout_us)
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:727]: comment says ms, vars say us - same in the header file

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 10]
```diff
@@ -0,0 +1,849 @@
...
+/**
+ * @details Tries to program as much as possible starting from the offset
+ *          until size. User needs to call this as many times as needed
+ *          taking already programmed bytes into account.
+ *
+ * @param[in] offset        address offset for program
+ * @param[in] data          data to be programmed
+ * @param[in] size          size desired to be programmed
+ * @param[out] programming  number of bytes programming, taking care of the limits
+ * @param[out] delay_ms     Time to wait until program typically finishes
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:751]: comment says ms,  vars say us

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+
+#include "AP_FlashIface_Abstract.h"
+
+
+class AP_FlashIface_JEDEC : public AP_FlashIface
+{
+public:
+    bool init() override;
+
+    // Read methods
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:30]: It's a comment but its not telling me anything useful :)

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    bool read(uint32_t offset, uint8_t *data, uint32_t size) override;
+
+
+    // Program methods
+    uint32_t get_page_size() override
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:35]: const

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        return _desc.page_size;
+    }
+    uint32_t get_page_count() override
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:39]: const + formatting

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        return _desc.page_count;
+    }
+    bool start_program_page(uint32_t page, uint8_t *data, uint32_t &delay_us, uint32_t &timeout_us) override;
+    bool start_program_offset(uint32_t offset, uint8_t* data, uint32_t size, uint32_t &programming,
+                              uint32_t &delay_us, uint32_t &timeout_us) override;
+
+    // Erase Methods
+    bool start_mass_erase(uint32_t &delay_ms, uint32_t &timeout_ms) override;
+    uint32_t get_sector_size() override
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:49]: const + formatting

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        return _desc.sector_size;
+    }
+    uint32_t get_sector_count() override
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:53]: const + formatting

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        return _desc.sector_count;
+    }
+    uint32_t min_erase_size() override
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:57]: const + formatting

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+    bool send_cmd(uint8_t ins, bool quad_mode);
+
+    bool _quad_spi_mode;
+
+private:
+    AP_HAL::OwnPtr<AP_HAL::QSPIDevice> _dev;
+
+    // Device description extracted from SFDP
+    struct device_desc {
+        uint16_t param_rev;
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:94]: JEDEC must doc these, can we insert the doc as comments?

**Review by andyp1per — COMMENTED** [review]
Half-way through

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.h (hunk 17 above)
Comment (bugobliterator) [reply libraries/AP_FlashIface/AP_FlashIface_JEDEC.h:94]: yes, i did, when we extract these variables in the code.

File: libraries/AP_HAL/QSPIDevice.h [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+#endif
+
+namespace AP_HAL
+{
+
+// Underlying HAL implementation can override these
+#if HAL_USE_QSPI_DEFAULT_CFG
+namespace QSPI
+{
+constexpr uint32_t CFG_CMD_MODE_MASK        =   (3LU << 8LU);
```
Comment (andyp1per) [inline libraries/AP_HAL/QSPIDevice.h:37]: This is redefing stuff that is already defined in chibios - why is that necessary?

File: libraries/AP_HAL/QSPIDevice.h [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+class QSPIDeviceManager
+{
+public:
+    virtual OwnPtr<QSPIDevice> get_device(const char *name)
+    {
+        return nullptr;
+    }
+
+    /* Return the number of QSPI devices currently registered. */
+    virtual uint8_t get_count()
```
Comment (andyp1per) [inline libraries/AP_HAL/QSPIDevice.h:113]: const + formatting

File: libraries/AP_HAL/QSPIDevice.h [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        return 0;
+    }
+
+    /* Get qspi device name at @idx */
+    virtual const char *get_device_name(uint8_t idx)
```
Comment (andyp1per) [inline libraries/AP_HAL/QSPIDevice.h:119]: const + formatting

File: libraries/AP_HAL/QSPIDevice.h [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+public:
+
+    QSPIDevice() : Device(BUS_TYPE_QSPI) { }
+
+    /* See AP_HAL::Device::transfer() */
+    virtual bool transfer(const uint8_t *send, uint32_t send_len,
+                          uint8_t *recv, uint32_t recv_len) override = 0;
+
+    // Set command header for upcomming transfer call(s)
+    virtual void set_cmd_header(CommandHeader cmd_hdr) override = 0;
```
Comment (andyp1per) [inline libraries/AP_HAL/QSPIDevice.h:96]: const CommandHeader&

File: libraries/AP_HAL_ChibiOS/QSPIDevice.cpp [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+    } else {
+        // Can't handle this transaction type
+        ret = false;
+    }
+    bus.bouncebuffer_finish(send, recv, recv_len);
+    acquire_bus(false);
+    return ret;
+}
+
+void QSPIDevice::set_cmd_header(CommandHeader cmd_hdr)
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/QSPIDevice.cpp:97]: const CommandHeader&

File: libraries/AP_HAL_ChibiOS/QSPIDevice.cpp [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+    if (!bus.semaphore.check_owner()) {
+        return false;
+    }
+    if (acquire) {
+        wspiAcquireBus(qspi_devices[device_desc.bus].driver);
+        if (qspi_devices[device_desc.bus].driver->config != &bus.wspicfg) {
+            // Initialise and Start WSPI driver
+            bus.wspicfg.end_cb = nullptr;
+            bus.wspicfg.error_cb = nullptr;
+            bus.wspicfg.dcr = STM32_DCR_FSIZE(24U) |        /* 16MB device.         */
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/QSPIDevice.cpp:114]: These both need to be configurable through hwdef.dat

File: libraries/AP_HAL_ChibiOS/QSPIDevice.h [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+};
+
+class QSPIBus : public DeviceBus
+{
+public:
+    QSPIBus(uint8_t _bus) :
+        DeviceBus(APM_SPI_PRIORITY, true),
+        bus(_bus) {}
+
+    uint8_t bus;
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/QSPIDevice.h:55]: private?

File: libraries/AP_HAL_ChibiOS/QSPIDevice.h [hunk 25]
```diff
@@ -0,0 +1,125 @@
...
+{
+
+struct QSPIDesc {
+    QSPIDesc(const char *_name, uint8_t _bus,
+             uint32_t _mode, uint32_t _speed)
+        : name(_name), bus(_bus), mode(_mode), speed(_speed)
+    {
+    }
+
+    const char *name;
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/QSPIDevice.h:41]: Some comments here would be good

**Review by andyp1per — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL/QSPIDevice.h (hunk 18 above)
Comment (bugobliterator) [reply libraries/AP_HAL/QSPIDevice.h:37]: So, the reason is we do not want to create dependency of Drivers running on QSPI on ChibiOS. There might be other HALs that might want to use those drivers.

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:90]: No GCS is bootloader. 😆 even console is not present, so this should actually be Debug a print

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 26]
```diff
@@ -0,0 +1,959 @@
...
+        _dev = hal.qspi->get_device(supported_devices[i].name);
+#endif
+        if (_dev) {
+            _dev_list_idx = i;
+            break;
+        }
+    }
+
+    if (!_dev) {
+        AP_HAL::panic("Ext Flash Not Found!");
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:115]: board config error maybe?

File: libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp [hunk 27]
```diff
@@ -0,0 +1,959 @@
...
+    } else {
+        return !(status & 1<<7);
+    }
+}
+
+// wait for the chip to be ready for the next instruction
+void AP_FlashIface_JEDEC::wait_ready()
+{
+    while (is_device_busy()) {
+        DELAY_MICROS(100);
```
Comment (andyp1per) [inline libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:881]: I think maybe this delay is too long - but I'm not sure what a good value is

Comment (andyp1per) [reply libraries/AP_FlashIface/AP_FlashIface_JEDEC.cpp:881]: Probably there should be a timeout and hard fail also

File: libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef.dat [hunk 28]
```diff
@@ -0,0 +1,88 @@
+# hw definition file for processing by chibios_hwdef.py
```
Comment (andyp1per) [inline libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef.dat:1]: It would probably also make sense to add a config for the DevEBox board

**Review by andyp1per — APPROVED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef-bl.dat [hunk 29]
```diff
@@ -0,0 +1,85 @@
...
+define CORE_CM7
+define SMPS_PWR
+
+define WATCHDOG_DISABLED
+
+# crystal frequency
+OSCILLATOR_HZ 25000000
+
+# board ID for firmware load
+APJ_BOARD_ID 1402
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef-bl.dat:21]: need to allocate this ID

File: libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef.dat [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+define CORE_CM7
+define SMPS_PWR
+
+define WATCHDOG_DISABLED
+
+# crystal frequency
+OSCILLATOR_HZ 25000000
+
+# board ID for firmware load
+APJ_BOARD_ID 1402
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/H757I_EVAL/hwdef.dat:21]: need to allocate an ID

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32f1_mcuconf.h [hunk 31]
```diff
@@ -26,6 +26,9 @@
  * 0...3        Lowest...Highest.
  */
 
+#define MHZ (1000U*1000U)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32f1_mcuconf.h:29]: emit in chibios_hwdef.py

**Review by tridge — CHANGES_REQUESTED** [review]

**Review by tridge — APPROVED** [review]
